extensions/panache/mongodb-panache/runtime/src/main/java/io/quarkus/mongodb/panache/reactive/runtime/ReactiveMongoOperations.java

```diff
@@ -299,36 +301,34 @@ private static ReactiveMongoCollection mongoCollection(Object entity) {
     }
 
     private static ReactiveMongoDatabase mongoDatabase(MongoEntity entity) {
-        ReactiveMongoClient mongoClient = mongoClient(entity);
+        ReactiveMongoClient mongoClient = clientFromArc(entity, ReactiveMongoClient.class);
         if (entity != null && !entity.database().isEmpty()) {
             return mongoClient.getDatabase(entity.database());
         }
-        String databaseName = getDefaultDatabaseName();
+        String databaseName = getDefaultDatabaseName(entity);
         return mongoClient.getDatabase(databaseName);
     }
 
-    private static String getDefaultDatabaseName() {
-        if (defaultDatabaseName == null) {
-            synchronized (MongoOperations.class) {
-                if (defaultDatabaseName == null) {
-                    defaultDatabaseName = ConfigProvider.getConfig()
-                            .getValue(MONGODB_DATABASE, String.class);
+    private static String getDefaultDatabaseName(MongoEntity entity) {
+        final String beanName = beanName(entity);
+        return defaultDatabaseName.computeIfAbsent(beanName, new Function<String, String>() {
+            @Override
+            public String apply(String s) {
+                MongoClients mongoClients = Arc.container().instance(MongoClients.class).get();
+                MongoClientConfig matchingMongoClientConfig = mongoClients.getMatchingMongoClientConfig(beanName);
+                if (matchingMongoClientConfig.database.isPresent()) {
+                    return matchingMongoClientConfig.database.get();
                 }
-            }
-        }
-        return defaultDatabaseName;
-    }
 
-    private static ReactiveMongoClient mongoClient(MongoEntity entity) {
-        if (entity != null && !entity.clientName().isEmpty()) {
-            Set<Bean<?>> beans = Arc.container().beanManager().getBeans(ReactiveMongoClient.class);
-            for (Bean<?> bean : beans) {
-                if (bean.getName() != null) {
-                    return (ReactiveMongoClient) Arc.container().instance(entity.clientName()).get();
+                MongoClientConfig defaultMongoClientConfig = mongoClients
+                        .getMatchingMongoClientConfig(MongoClientBeanUtil.DEFAULT_MONGOCLIENT_NAME);
+                if (defaultMongoClientConfig.database.isPresent()) {
+                    return defaultMongoClientConfig.database.get();
                 }
+
+                throw new IllegalArgumentException("The database property was not configured for " + entity);
```

Comment:
I think we need to better explain what went wrong as the rules are a little complex. And don't forget that `entity` here can be null so at least you need to provides different error message in case an `entity` configuration exist (so you can add the name of the entity inside the error message) and none are set.

Normally, if we are here, this means that no `database` is configured via the entity annotation so one must be set inside the application.properties, we shoud compute the needed property and log it.

Reply:
Makes sense, I'll try and make it clearer

Reply:
What do you think about it now?